### PR TITLE
Add metrics for authentication config reload

### PR DESCRIPTION
**Pull request by enj**

### CODE REVIEW

File: pkg/kubeapiserver/options/authentication.go [hunk 1]
```diff
@@ -676,15 +677,18 @@ func (o *BuiltInAuthenticationOptions) ApplyTo(ctx context.Context, authInfo *ge
 					klog.ErrorS(err, "failed to validate authentication config")
 					// this config is not structurally valid and never will be, update the tracker so we stop retrying
 					trackedAuthenticationConfigData = authConfigData
+					authenticationconfigmetrics.RecordAuthenticationConfigAutomaticReloadFailure(apiServerID)
```
Comment (enj) [inline pkg/kubeapiserver/options/authentication.go:705]: Don't we need this for the load above too?  Maybe this should be done via a `defer` so it is easier to track?

File: staging/src/k8s.io/apiserver/pkg/server/options/authenticationconfig/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+			Name:           "automatic_reload_last_timestamp_seconds",
+			Help:           "Timestamp of the last automatic reload of authentication configuration split by status and apiserver identity.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"status", "apiserver_id_hash"},
+	)
+)
+
+var registerMetrics sync.Once
+var hashPool *sync.Pool
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/authenticationconfig/metrics/metrics.go]: Lets drop all the `hashPool` stuff, we have never actually seen any performance justification to do this (and probably should undo this pattern elsewhere).

File: staging/src/k8s.io/apiserver/pkg/server/options/authenticationconfig/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+			New: func() interface{} {
+				return sha256.New()
+			},
+		}
+		legacyregistry.MustRegister(authenticationConfigAutomaticReloadsTotal)
+		legacyregistry.MustRegister(authenticationConfigAutomaticReloadLastTimestampSeconds)
+	})
+}
+
+func ResetMetricsForTest() {
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/authenticationconfig/metrics/metrics.go:66]: Do we need to do this `Reset` right before the swap of the new authenticator?

File: test/integration/apiserver/oidc/oidc_test.go [hunk 4]
```diff
@@ -1275,11 +1300,18 @@ jwt:
 				Groups:   []string{"system:authenticated"},
 			},
 			waitAfterConfigSwap: true,
+			wantMetricStrings: []string{
+				`apiserver_authentication_config_controller_automatic_reload_last_timestamp_seconds{apiserver_id_hash="sha256:3c607df3b2bf22c9d9f01d5314b4bbf411c48ef43ff44ff29b1d55b41367c795",status="failure"} FP`,
+				`apiserver_authentication_config_controller_automatic_reloads_total{apiserver_id_hash="sha256:3c607df3b2bf22c9d9f01d5314b4bbf411c48ef43ff44ff29b1d55b41367c795",status="failure"} 1`,
```
Comment (enj) [inline test/integration/apiserver/oidc/oidc_test.go:1307]: Are we sure this exact counter won't flake on CI?  i.e. could it be `2`?

**Review by enj — COMMENTED** [review]
Minor stuff.

File: staging/src/k8s.io/apiserver/pkg/server/options/authenticationconfig/metrics/metrics.go (hunk 3 above)
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/server/options/authenticationconfig/metrics/metrics.go:66]: Never mind on this, I am thinking about the JWKs metrics that are specific to issuer.

File: pkg/kubeapiserver/options/authentication.go (hunk 1 above)
Comment (liggitt) [reply pkg/kubeapiserver/options/authentication.go:705]: ha, I suggested it not be a defer so it was obvious all the places we trip a failure metric ... I only care about 3 out of 10 but I found the defer harder to reason about (as well as counting on `err` not being masked through the whole function)

**Pull request closed without merging**